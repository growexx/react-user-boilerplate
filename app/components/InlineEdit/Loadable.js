/**
 *
 * Asynchronously loads the component for InlineEdit
 *
 */

import loadable from 'utils/loadable';

export default loadable(() => import('./index'));
