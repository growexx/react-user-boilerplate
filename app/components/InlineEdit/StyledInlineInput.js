import styled from 'styled-components';

export const StyledInlineInput = styled.div`
  .inputValue {
    height: auto;
    padding: 8px;
  }
  .inputValue:hover {
    cursor: text;
    background: @backgroundGrey;
  }
  .placeHolderView {
    color: @grey !important;
  }
`;
