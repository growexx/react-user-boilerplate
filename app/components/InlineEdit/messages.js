/*
 * InlineEdit Messages
 *
 * This contains all the text for the InlineEdit component.
 */

import { defineMessages } from 'react-intl';

export const scope = 'app.components.InlineEdit';

export default defineMessages({
  inputValue: {
    id: `${scope}.inputValue`,
    defaultMessage: `{inputValue}`,
  },
});
