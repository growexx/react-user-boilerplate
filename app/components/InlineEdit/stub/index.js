export const TEST_IDS = {
  INPUT_VALUE: 'INPUT_VALUE',
  INPUT_EDIT: 'INPUT_EDIT',
  SAVE_BUTTON: 'save',
  CANCEL_BUTTON: 'cancel',
};
