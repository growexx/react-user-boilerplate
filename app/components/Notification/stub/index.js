import React from 'react';
import {
  UserOutlined,
  CheckCircleOutlined,
  SmileOutlined,
  ExportOutlined,
  NumberOutlined,
} from '@ant-design/icons';
const NOTIFICATIONS = [
  {
    icon: <UserOutlined />,
    update: '70 new employees are shifted',
    timestamp: 1596119688264,
    read: false,
  },
  {
    icon: <CheckCircleOutlined />,
    update: 'Time to Take a Break, TADA!!!',
    timestamp: 1596119686811,
    read: false,
  },
  {
    icon: <SmileOutlined />,
    update: 'Time to Take a Break, TADA!!!',
    timestamp: 1596119686811,
    read: false,
  },
  {
    icon: <ExportOutlined />,
    update: 'Time to Take a Break, TADA!!!',
    timestamp: 1596119686811,
    read: false,
  },
  {
    icon: <NumberOutlined />,
    update: '70 new employees are shifted',
    timestamp: 1596119688264,
    read: false,
  },
];

export const TEST_IDS = {
  MARK_ALL_READ: 'Mark All As Read',
  BELL_ICON: 'BELL_ICON',
  EMPTY_CONTAINER: 'EMPTY_CONTAINER',
  INFINITE_SCROLLING: 'INFINITE_SCROLLING',
  NOTIFICATION_ITEM: 'NOTIFICATION_ITEM',
};

export const getNotificationsSuccessMock = () =>
  Promise.resolve({
    data: NOTIFICATIONS,
    status: 1,
  });
export const getNotificationsMockWithNoData = () =>
  Promise.resolve({
    data: [],
    status: 0,
  });
export const getNotificationsMockWithLessData = () =>
  Promise.resolve({
    data: [NOTIFICATIONS[0]],
    status: 1,
  });
export const getNotificationsFailureData = () =>
  Promise.reject(
    new Error({
      message: 'something went wrong',
    }),
  );

export const getNotificationsSuccessMockForClick = () =>
  Promise.resolve({
    data: [
      {
        icon:
          'https://encrypted-tbn0.gstatic.com/images?q=tbn:ANd9GcQZZrv3_PEnkdOIZvnr0COONt3kL7rSSq623dB3fyLCgT7GARpReF26nPOre6JCLHKu7KQ&usqp=CAU',
        update: '70 new employees are shifted',
        timestamp: 1596119688264,
        read: false,
        clickAction: 'http://localhost:3000/features',
      },
      {
        icon: <UserOutlined />,
        update: '70 new employees are shifted',
        timestamp: 1596119688264,
        read: true,
      },
      ...NOTIFICATIONS,
    ],
    status: 1,
  });
