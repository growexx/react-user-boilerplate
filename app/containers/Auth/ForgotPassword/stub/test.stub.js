export const TEST_IDS = {
  RESET_PASSWORD: 'RESET_PASSWORD',
};
