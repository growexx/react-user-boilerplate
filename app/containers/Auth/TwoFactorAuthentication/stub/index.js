export const stubUserData = elseCase => {
  if (elseCase) {
    return {
      docs: [],
    };
  }
  return {
    docs: [
      {
        id: '0',
      },
    ],
  };
};
