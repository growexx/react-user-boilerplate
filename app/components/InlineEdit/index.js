/**
 *
 * InlineEdit
 *
 */

import React, { memo, createRef } from 'react';
import { Input, Button } from 'antd';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCheck, faTimes } from '@fortawesome/free-solid-svg-icons';
import PropTypes from 'prop-types';
import { ESC_KEY_CODE } from 'components/InlineEdit/constants';
import { StyledInlineInput } from 'components/InlineEdit/StyledInlineInput';
import { TEST_IDS } from 'components/InlineEdit/stub';
class InlineEdit extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      isInputActive: false,
      inputValue: props.value,
    };
    this.inputContainerRef = createRef(null);
  }

  handleOutSide = e => {
    const { isInputActive } = this.state;
    if (
      this.inputContainerRef &&
      this.inputContainerRef.current &&
      this.inputContainerRef.current.contains(e.target)
    ) {
      return;
    }
    if (isInputActive) {
      this.onSave();
    }
  };

  componentDidMount() {
    document.addEventListener('mousedown', this.handleOutSide);
    document.addEventListener('touchstart', this.handleOutSide);
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleOutSide);
    document.removeEventListener('touchstart', this.handleOutSide);
  }

  onSave = () => {
    const { inputValue } = this.state;
    const { onSave } = this.props;
    onSave(inputValue);
    this.setState({
      isInputActive: false,
    });
  };

  onCancel = () => {
    const { value } = this.props;
    this.setState({
      inputValue: value,
      isInputActive: false,
    });
  };

  handleKeyDown = e => {
    if (e.keyCode === ESC_KEY_CODE) {
      this.onCancel();
    }
  };

  handleOnChange = e => {
    this.setState({
      inputValue: e.target.value,
    });
  };

  handleDoubleClick = () => {
    this.setState({
      isInputActive: true,
    });
  };

  saveIcon = () => (
    <Button type="link">
      <FontAwesomeIcon
        icon={faCheck}
        onClick={this.onSave}
        data-testid={TEST_IDS.SAVE_BUTTON}
        title={TEST_IDS.SAVE_BUTTON}
      />
    </Button>
  );

  cancelIcon = () => (
    <Button type="link">
      <FontAwesomeIcon
        icon={faTimes}
        onClick={this.onCancel}
        data-testid={TEST_IDS.CANCEL_BUTTON}
        title={TEST_IDS.CANCEL_BUTTON}
      />
    </Button>
  );

  addonAfterView = () => (
    <span>
      {this.saveIcon()}
      {this.cancelIcon()}
    </span>
  );

  getValue = () => {
    const { value, placeholder } = this.props;
    if (value.length > 0) {
      return value;
    }
    return <span className="placeHolderView">{placeholder}</span>;
  };

  render() {
    const { inputValue, isInputActive } = this.state;
    const { placeholder } = this.props;
    return (
      <StyledInlineInput ref={this.inputContainerRef}>
        {!isInputActive ? (
          <div
            data-testid={TEST_IDS.INPUT_VALUE}
            onDoubleClick={this.handleDoubleClick}
            className="inputValue"
          >
            {this.getValue()}
          </div>
        ) : (
          <Input
            placeholder={placeholder}
            suffix={this.addonAfterView()}
            value={inputValue}
            onChange={this.handleOnChange}
            allowClear
            onPressEnter={this.onSave}
            onKeyDown={this.handleKeyDown}
            data-testid={TEST_IDS.INPUT_EDIT}
          />
        )}
      </StyledInlineInput>
    );
  }
}

InlineEdit.propTypes = {
  value: PropTypes.string,
  onSave: PropTypes.func,
  placeholder: PropTypes.string,
};

export default memo(InlineEdit);
