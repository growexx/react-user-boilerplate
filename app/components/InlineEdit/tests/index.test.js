/**
 *
 * Tests for InlineEdit
 *
 * @see https://github.com/react-boilerplate/react-boilerplate/tree/master/docs/testing
 *
 */

import React from 'react';
import { fireEvent, render } from 'react-testing-library';
import userEvent from '@testing-library/user-event';
import { IntlProvider } from 'react-intl';
import 'jest-dom/extend-expect';
import { Provider } from 'react-redux';
import { ConnectedRouter } from 'connected-react-router';
import { browserHistory } from 'react-router-dom';
import history from 'utils/history';
import configureStore from 'configureStore';
import InlineEdit from 'components/InlineEdit/index';
import { TEST_IDS } from 'components/InlineEdit/stub';
let store;
const props = {
  onSave: jest.fn(),
  value: 'testValue',
  placeholder: 'This is test component',
};
const componentWrapper = () =>
  render(
    <Provider store={store}>
      <IntlProvider locale="en">
        <ConnectedRouter history={history}>
          <InlineEdit {...props} />
        </ConnectedRouter>
      </IntlProvider>
    </Provider>,
  );

describe('<InlineEdit />', () => {
  beforeAll(() => {
    store = configureStore({}, browserHistory);
  });
  it('Should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = componentWrapper();
    expect(firstChild).toMatchSnapshot();
  });
  it('Should double click and input box is appeared', () => {
    const { getByTestId } = componentWrapper();
    // show the input
    userEvent.dblClick(getByTestId(TEST_IDS.INPUT_VALUE));
    expect(getByTestId(TEST_IDS.INPUT_EDIT)).toBeInTheDocument();
  });
  it('Should double click and cancel the editing', () => {
    const { getByTestId, getByText } = componentWrapper();
    // show the input
    userEvent.dblClick(getByTestId(TEST_IDS.INPUT_VALUE));
    expect(getByTestId(TEST_IDS.INPUT_EDIT)).toBeInTheDocument();
    fireEvent.change(getByTestId(TEST_IDS.INPUT_EDIT), {
      target: {
        value: 'testInputChanged',
      },
    });
    // cancel the input
    fireEvent.click(getByTestId(TEST_IDS.CANCEL_BUTTON));
    expect(getByText('testValue')).toBeInTheDocument();
  });
  it('Should double click and write in input box', async () => {
    const { getByTestId } = componentWrapper();
    // show the input
    userEvent.dblClick(getByTestId(TEST_IDS.INPUT_VALUE));
    expect(getByTestId(TEST_IDS.INPUT_EDIT)).toBeInTheDocument();
    // write on the input
    fireEvent.change(getByTestId(TEST_IDS.INPUT_EDIT), {
      target: {
        value: 'testInputChanged',
      },
    });
    // save the input
    fireEvent.click(getByTestId(TEST_IDS.SAVE_BUTTON));
    expect(getByTestId(TEST_IDS.INPUT_VALUE)).toBeInTheDocument();
  });
  it('Should double click and press escape', () => {
    props.value = '';
    const { getByTestId, queryByTestId } = componentWrapper();
    // show the input
    userEvent.dblClick(getByTestId(TEST_IDS.INPUT_VALUE));
    expect(getByTestId(TEST_IDS.INPUT_EDIT)).toBeInTheDocument();
    fireEvent.keyDown(getByTestId(TEST_IDS.INPUT_EDIT), {
      keyCode: 27,
    });
    expect(queryByTestId(TEST_IDS.INPUT_VALUE)).toBeInTheDocument();
  });
  it('Should double click and press enter button', () => {
    const { getByTestId, queryByTestId } = componentWrapper();
    // show the input
    userEvent.dblClick(getByTestId(TEST_IDS.INPUT_VALUE));
    expect(getByTestId(TEST_IDS.INPUT_EDIT)).toBeInTheDocument();
    fireEvent.keyDown(getByTestId(TEST_IDS.INPUT_EDIT), {
      keyCode: 13,
    });
    expect(queryByTestId(TEST_IDS.INPUT_VALUE)).toBeInTheDocument();
  });
  it('Should double click and fire mouse down', () => {
    const { getByTestId } = componentWrapper();
    // show the input
    userEvent.dblClick(getByTestId(TEST_IDS.INPUT_VALUE));
    // click on input should not close it
    fireEvent.mouseDown(getByTestId(TEST_IDS.INPUT_EDIT));
    expect(getByTestId(TEST_IDS.INPUT_EDIT)).toBeInTheDocument();
    // click on outside input should close it
    fireEvent.click(getByTestId(TEST_IDS.CANCEL_BUTTON));
    expect(getByTestId(TEST_IDS.INPUT_VALUE)).toBeInTheDocument();
  });
});
