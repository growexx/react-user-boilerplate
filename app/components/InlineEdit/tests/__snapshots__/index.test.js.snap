// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InlineEdit /> Should render and match the snapshot 1`] = `
.c0 .inputValue {
  height: auto;
  padding: 8px;
}

.c0 .inputValue:hover {
  cursor: text;
  background: rgba(9,30,66,0.04);
}

.c0 .placeHolderView {
  color: #808080 !important;
}

<div
  class="c0"
>
  <div
    class="inputValue"
    data-testid="INPUT_VALUE"
  >
    testValue
  </div>
</div>
`;
